llvm/lib/Transforms/IPO/SampleProfile.cpp

```diff
@@ -2443,53 +2374,232 @@ void SampleProfileMatcher::runOnFunction(const Function &F) {
   }
 }
 
-void SampleProfileMatcher::runOnModule() {
-  ProfileConverter::flattenProfile(Reader.getProfiles(), FlattenedProfiles,
-                                   FunctionSamples::ProfileIsCS);
-  for (auto &F : M) {
-    if (F.isDeclaration() || !F.hasFnAttribute("use-sample-profile"))
+void SampleProfileMatcher::recordCallsiteMatchStates(
+    const Function &F, const std::map<LineLocation, StringRef> &IRAnchors,
+    const std::map<LineLocation, std::unordered_set<FunctionId>>
+        &ProfileAnchors,
+    const LocToLocMap *IRToProfileLocationMap) {
+  bool IsPostMatch = IRToProfileLocationMap != nullptr;
+  auto &CallsiteMatchStates =
+      FuncCallsiteMatchStates[FunctionSamples::getCanonicalFnName(F.getName())];
+
+  auto MapIRLocToProfileLoc = [&](const LineLocation &IRLoc) {
+    // IRToProfileLocationMap is null in pre-match phrase.
+    if (!IRToProfileLocationMap)
+      return IRLoc;
+    const auto &ProfileLoc = IRToProfileLocationMap->find(IRLoc);
+    if (ProfileLoc != IRToProfileLocationMap->end())
+      return ProfileLoc->second;
+    else
+      return IRLoc;
+  };
+
+  for (const auto &I : IRAnchors) {
+    // In post-match, use the matching result to remap the current IR callsite.
+    const auto &ProfileLoc = MapIRLocToProfileLoc(I.first);
+    const auto &IRCalleeName = I.second;
+    const auto &It = ProfileAnchors.find(ProfileLoc);
+    if (It == ProfileAnchors.end())
       continue;
-    runOnFunction(F);
+    const auto &Callees = It->second;
+
+    bool IsCallsiteMatched = false;
+    // Since indirect call does not have CalleeName, check conservatively if
+    // callsite in the profile is a callsite location. This is to reduce num of
+    // false positive since otherwise all the indirect call samples will be
+    // reported as mismatching.
+    if (IRCalleeName == SampleProfileMatcher::UnknownIndirectCallee)
+      IsCallsiteMatched = true;
+    else if (Callees.size() == 1 && Callees.count(getRepInFormat(IRCalleeName)))
+      IsCallsiteMatched = true;
+
+    if (IsCallsiteMatched) {
+      auto R =
+          CallsiteMatchStates.emplace(ProfileLoc, MatchState::InitialMatch);
+      // When there is an existing state, we know it's in post-match phrase.
+      // Update the matching state accordingly.
+      if (!R.second && R.first->second == MatchState::InitialMismatch)
+        R.first->second = MatchState::RecoveredMismatch;
+    }
+  }
+
+  // Check if there are any callsites in the profile that does not match to any
+  // IR callsites.
+  for (const auto &I : ProfileAnchors) {
+    const auto &Loc = I.first;
+    [[maybe_unused]] const auto &Callees = I.second;
+    assert(!Callees.empty() && "Callees should not be empty");
+    auto It = CallsiteMatchStates.find(Loc);
+    if (It == CallsiteMatchStates.end())
+      CallsiteMatchStates.emplace(Loc, MatchState::InitialMismatch);
+    // The inital match is removed in post-match.
+    else if (IsPostMatch && It->second == MatchState::InitialMatch)
```

Comment:
A bit confused here, if there was an entry and it was matched in pre-match, it won't be changed in post-match if it still matches, why do we change it to RemovedMatch?

Reply:
Oops, Good catch, this is wrong here, it's update all InitialMatch to RemovedMatch regardless it's matched or not. 

The fact is that we don't have a `FinalMatch` state in current implementation, so we are not able to determine whether it's matched or not in post-match in this loop. 

But we also can't only use the `FinalMatch` and `FinalMismatch`, because the counting is at the end of the function, then we don't know the recovered numbers(which is this diff's goal)

It think it's clear that in pre-match we can have 'InitialMatch' and 'IntialMismatch'

but in post match, I think we need 4 states, but that make it more complicated.. 
1)  'InitialMatch'  stay match,    to `UnchangedMatch`
2)  'IntialMismatch'  stay mismatch,    to `UnchangedMismatch`
3)  'IntialMismatch' is recovered.  to `RecoveredMismatch`
4)  'InitialMatch'  is removed.  to `RemovedMatch`

Reply:
So in total there would be 6 states, the  `InitialMatch ` and `InitialMismatch ` is only used for transition. In the `counting` function, we only do counting on the 4 post-match states. `UnchangedMismatch` + `RecoveredMismatch ` is the original mismatch,`RemovedMatch` + `UnchangedMismatch ` is the final mismatch,  `RecoveredMismatch ` is the recovered mismatch.  

Any thoughts?

Reply:
6 explicit state sounds ok with me, that should make things really clear? 

I'd add comment for each of the state.

Reply:
Also the fact that this was wrong and tests all passed means that the tests are probably not strong enough.

Reply:
>6 explicit state sounds ok with me, that should make things really clear?
yeah, I feel it's clearer, please take another look.

>Also the fact that this was wrong and tests all passed means that the tests are probably not strong enough.
Uh, I see, I checked that it's due to the tests are ether all matched cases or all mismatched cases. Updated the test.